llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1198,24 +1198,37 @@ Value *SPIRVToLLVM::transConvertInst(SPIRVValue *bv, Function *f, BasicBlock *bb
 void SPIRVToLLVM::setFastMathFlags(SPIRVValue *bv) {
   // For floating-point operations, if "FastMath" is enabled, set the "FastMath"
   // flags on the handled instruction
+  if (!SPIRVGenFastMath)
+    return;
+
+  // Only do this for operations with floating point type.
+  if (!bv->hasType())
+    return;
+  SPIRVType *ty = bv->getType();
+  if (ty->isTypeVector())
+    ty = ty->getVectorComponentType();
+  if (!ty->isTypeFloat())
+    return;
+
   llvm::FastMathFlags fmf;
-  if (SPIRVGenFastMath) {
-    fmf.setAllowReciprocal();
-    // Enable contraction when "NoContraction" decoration is not specified
-    bool allowContract = !bv || !bv->hasDecorate(DecorationNoContraction);
-    // Do not set AllowContract or AllowReassoc if DenormFlushToZero is on, to
-    // avoid an FP operation being simplified to a move that does not flush
-    // denorms.
-    if (m_fpControlFlags.DenormFlushToZero == 0) {
-      fmf.setAllowContract(allowContract);
-      // AllowRessociation should be same with AllowContract
-      fmf.setAllowReassoc(allowContract);
-    }
-    // Enable "no NaN" and "no signed zeros" only if there isn't any floating point control flags
-    if (m_fpControlFlags.U32All == 0) {
-      fmf.setNoNaNs();
-      fmf.setNoSignedZeros(allowContract);
-    }
+  fmf.setAllowReciprocal();
+  if (ty->isTypeFloat(32))
```

Comment:
This also does not allow afn for half-float. Is that intended? (Also brace this as it is multi-line including the comment.)

Reply:
We don't have fp16 transcendental operations anyway. Since approximate fp32 versions of transcendentals should have sufficient precision when applied to fp16 inputs, we should ensure that any backend lowering to fp32 applies the appropriate flags, but I think the right place to do that is in LLVM rather than here in the frontend.

Reply:
Huh, actually that's wrong -- we do have fp16 assembler instructions for some transcendentals,
but I believe those are of sufficient accuracy anyway.